### DISCUSSION TITLE
Suggestion: Style scoping with pure CSS

### DISCUSSION THREAD

**Discussion by dojyorin** [vuejs/rfcs#707] [General]
Currently, scoped CSS is done by compiler.
However, today's CSS has great [@scope](https://developer.mozilla.org/docs/Web/CSS/@scope) rule.
By using this, I think we can achieve CSS scoping even more simply.

Example 1
Currently, if include `<style>` inside `<template>` it will be ignored, but if only `@scope` exists it will be passed through.

```vue
<!-- Good -->
<template>
    <style>
        @scope {
            .foo {
                background-color: cyan;
            }
        }
    </style>
    <div class="foo">Scoped</div>
</template>

<!-- Bad, ignored -->
<template>
    <style>
        .foo {
            background-color: cyan;
        }
    </style>

    <div class="foo">Scoped</div>
</template>
```

Example 2
Add `.css` or `.style` to component object.
Write your CSS rules inside it.
It's debatable whether to explicitly state `@scope` or add it internally.
Then insert contents of `.css/.style` directly under template root.

```js
// When to specify
export const MyComponent = defineComponent({
    template: `<div class="foo">Scoped</div>`,
    setup() {/*...*/},
    style: `
        @scope {
            .foo {
                background-color: cyan;
            }
        }
    `
});

// When automatically assigned internally
export const MyComponent = defineComponent({
    template: `<div class="foo">Scoped</div>`,
    setup() {/*...*/},
    style: `
        .foo {
            background-color: cyan;
        }
    `
});
```

I don't know details of how template engine works internally, but I think it goes something like this.

```js
// When to specify
const styleMergedTemplate = `<style>${component.style}</style>${component.template}`;
// When automatically assigned internally
const styleMergedTemplate = `<style>@scope{${component.style}}</style>${component.template}`;
```

Both will eventually be inserted into template.

```vue
<template>
    <style>
        @scope {
            .foo {
                background-color: cyan;
            }
        }
    </style>
    <div class="foo">Scoped</div>
</template>
```

**Note**

According to [Can I Use](https://caniuse.com/css-cascade-scope) FireFox is only modern browser that doesn't support `@scope`, but they are currently [working](https://bugzilla.mozilla.org/show_bug.cgi?id=1830512) on adding feature.

**Comment by jcupps** [comment]
Yes, or at least support targeting the component's root element(s) using [:scope](https://developer.mozilla.org/en-US/docs/Web/CSS/@scope#the_scope_pseudo-class), e.g.

```
<style scoped>
:scope { ... }
</style>
```

This would avoid the need to use arbitrary class names to target the root element, which I personally encounter frequently.

A similar proposal has been discussed here: https://github.com/vuejs/vue-loader/issues/1601

**Comment by webgoto** [comment]
I would also love this feature.
